modules/compute/vm-instance/variables.tf

```diff
@@ -195,11 +195,13 @@ variable "tags" {
 variable "threads_per_core" {
   description = <<-EOT
   Sets the number of threads per physical core. By setting threads_per_core
-  greater than 1, Simultaneous Multithreading (SMT) is enabled extending the
-  total number of virtual cores. For example, a machine of type c2-standard-60
-  will have 60 virtual cores with threads_per_core equal to 2. With
-  threads_per_core equal to 1 (SMT turned off), only the 30 physical cores will
-  be available on the VM.
+  to 2, Simultaneous Multithreading (SMT) is enabled extending the total number
+  of virtual cores. For example, a machine of type c2-standard-60 will have 60
+  virtual cores with threads_per_core equal to 2. With threads_per_core equal
+  to 1 (SMT turned off), only the 30 physical cores will be available on the VM.
+
+  The setting will be silently ignored for machine types that do not support 
+  SMT (t2d, shared-core instances, or instances with less than 2 vCPU).
```

Comment:
I definitely prefer `null` as the default value and hadn't considered the simple

```hcl
locals {
...
  threads_per_core = local.smt_capable && var.threads_per_core == null ? 1 : var.threads_per_core
...
} 
```

With the obvious substitution in the `for_each` block. Otherwise code (I think) remains the same.

Reply:
>I prefer that we avoid silently ignoring values whenever possible.

Agreed. I added a precondition on the vm_instance resource to make the manifest fail when threads_per_core is set on an unsupported instance. Here is an example error:

```
│ Error: Resource precondition failed
│ 
│   on modules/vm-instance/main.tf line 157, in resource "google_compute_instance" "compute_vm":
│  157:       condition     = var.threads_per_core == null || local.smt_capable
│     ├────────────────
│     │ local.smt_capable is false
│     │ var.threads_per_core is 2
│ 
│ machine type n1-standard-1 does not support threads_per_core setting on instance n1-1-thread-0
```

> I would suggest that we set the default value to null in [variables.tf](http://variables.tf/)

I considered this, but noticed that _(while unrelated to 'ghpc')_ the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/configure-smt) mentions 'null' as being the Compute Engine VM default. 
I decided against altering the behavior of `null` for consistency. Let me know if you think otherwise. 
If we are altering the default behavior of GCE, I think we should be explicit about it in our default vars.

> Then in the locals block we look for a value of null and replace with an intelligent default (1 where SMT off is supported, 2 everywhere else).

It would be "1 where SMT off is supported, and no config block everywhere else". We cannot have a `threads_per_core` parameter on instances where the SMT parameter is not supported, regardless of the value.

Reply:
Whatever version the final PR requires must also be accurately reflected in `modules/compute/vm-instance/versions.tf`.

Reply:
> If the user explicitly sets the value (var.threads_per_core != null), then we don't do any validation and pass through the users value regardless of correctness.

I was trying to avoid pass-through of the users value regardless of correctness, because this would only fail at the `terraform apply` stage, and would not be caught at `terraform plan`, however, after giving this a bit more thought, I believe this is not really an issue.

The behavior I have in mind is very similar to what @nick-stroud has described, with minor changes:

- in variables.tf keep default `threads_per_core = 1`
- ignore the smt parameter block if `threads_per_core == null`
- pass through whatever value the user provided if `threads_per_core != null`
 and let the GCE resource fail if an unsupported parameter is provided

With this, we do not need to make any assertions as to whether an machine type supports SMT or not, since we rely on the GCE API to fail, which makes us compatible with future instance types, without having to modify our logic.

And, we still give the user possibility to overwrite our explicit default value of `threads_per_core = 1` to `null` which will remove the config block for unsupported machine types.